### DISCUSSION TITLE
[Question] How to serialize query results?

### DISCUSSION THREAD

**Discussion by hugotox** [Vincit/objection.js#2515] [Questions]
Hello! 

This might be an issue related to Svelte, but I still think is worth asking the difference, or why a `knex` query result is serializable and an `objection` is not.

Example on a `+page.server.ts` in Svelte:

```typescript
import { Model } from 'objection'
import knex from 'knex'

const pg = knex({
  client: 'pg',
  connection: process.env.DATABASE_URL,
  searchPath: ['public'],
})

Model.knex(pg)

class User extends Model {
  static get tableName() {
    return 'users'
  }
}

export const load = async () => {
  const users = await User.query()
  return {
    users,
  }
}

```

The above code fails on Svelte with a `Cannot stringify arbitrary non-POJOs` error. I can circumvent this issue with `users: users.map((u) => u.toJSON()),` but seems inefficient for large resultsets.

What is most curious to me, is that a `knex` query works without parsing to JSON:

```typescript
import knex from 'knex'

const pg = knex({
  client: 'pg',
  connection: process.env.DATABASE_URL,
  searchPath: ['public'],
})

export const load = async () => {
  const rawUsers = await pg.select('*').from('users')
  return {
    rawUsers,
  }
}

```

Any advice on how to properly use Objection in Svelte is highly appreciated 🙏

**Comment by yayza** [comment]
That's pretty much the most straightforward and simplest way to do it, it's how I do it using SvelteKit. Since model queries return models as result we'll have to transform them to JSON. 

If you're working with large dataset, then I would suggest paginating results or streaming.

If you want to abstract away always using the map method on results you can create a custom QueryBuilder and add a new method `.toJson()` to it and chain that to your query:

```ts
export class CustomQueryBuilder<M extends Model, R> extends Model.QueryBuilder<M, R> {
    toJson() {
        return this.runAfter((result: any) => {
            if (Array.isArray(result)) {
                return result.map((r) => r.toJSON());
            } else if (result) {
                return result.toJSON();
            }
            return result;
        });
    }
}
```


```ts
export class BaseModel extends Model {
    static idColumn: string | string[] = "id";

    id: number;
    created_at: string;
    updated_at: string;

    static get QueryBuilder() {
        return CustomQueryBuilder;
    }
```
And you can use it like:
 ```ts
const queryResultsInJson = await BaseModel.query().toJson();
````
